UiaOperationAbstraction: expose GetPropertyValue as a method on UiaElement

michaelDCurran opened the pull request:
UiaOperationAbstraction's UiaElement class currently has getter methods for most if not all standard UIA properties (E.g. GetName, GetControlType etc). However, there is no way to cleanly fetch a property value by a given property ID. Exposing GetPropertyValue would be useful for the following reasons:
* Calling logic may want to fetch a long list of properties. Rather than having to hard-code calls to the individual getter methods, it would be much easier to just walk through a for loop of property IDs, calling GetPropertyValue generically with each property ID.
* UI Automation has the concept of custom properties, and Microsoft products seem to be moving towards using these more heavily these days. Assuming that Microsoft.UI.UIAutomation.AutomationRemoteElement.GetPropertyValue and its underlying machinary do already correctly marshal custom property IDs, then exposing GetProprtyValue on UiaElement would allow for fetching these.

This PR exposes GetPropertyValue on UiaElement. Remotely it uses AutomationRemoteElement.GetPropertyValue, and locally it uses IUIAutomationElement.GetCurrentPropertyValue or IUIAutomationElement.GetCachedPropertyValue (depending on whether caching is requested). It also takes a 'ignoreDefault' boolean argument, and its return type is a UiaVariant.
I have also added a basic functional test which fetches name via GetPropertyValue and asserts the string returned is correct.

zhayuli commented:
/azp run

beweedon commented on src/UIAutomation/UiaOperationAbstraction/UiaTypeAbstractionImpl.g.cpp:
```diff
@@ -5114,6 +5114,31 @@
         return localPropertyValue;
     }
 
+    UiaVariant UiaElement::GetPropertyValue(UiaPropertyId propId, UiaBool ignoreDefault, bool useCachedApi)
```
Add comment here specifying the default values of `ignoreDefault` and `useCachedApi`? (see `UiaElement::GetProcessId` below).

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,37 @@ namespace UiaOperationAbstractionTests
+            auto scope = UiaOperationScope::StartNew();
+
+            UiaElement element = calc;
+            auto val = element.GetPropertyValue(UiaPropertyId(UIA_NamePropertyId));
```
Add tests verifying `ignoreDefalut` and `useCachedApi` work as expected?

michaelDCurran replied:
I agree there should be tests, but I'm a little bit lost as to how to do this appropriately at the moment.
Testing ignoreDefault=false is easy enough. We just need to fetch a property we know is not implemented, and assert that it is its correct default empty value.
For ignoreDefault=true, we really want to test that fetching a property that is not implemented fails with an exception. Locally, I would assume this test framework has some kind of assertRaisesException, we could wrap around the call to GetPropertyValue, and ensure it does raise an exception. However, remotely, we don't see the exception until scope.resolve is called. 
For useCachedAPI: again locally, we could try fetching a property we know is not already cached, and test for the exception. Remotely, again we won't see this until scope.Resolve.
I'm aware that the lower level remote ops machinary now has support for try blocks. If there are plans to implement try blocks for the higher-level UiaOperationAbstraction, then I think it might be better to wait until that is there, and write the tests using that, which means they can be written the same way for both remote and local.

beweedon commented:
@mavangur @vinooc-ms Does this look good from the custom property/extensibility perspective?

mavangur commented:
@beweedon Looks good to me. I don't see any issues from extensibility side. I don't have practical knowledge on Property Extensibility, (worked only on Custom Annotations, Custom Patterns)but as far as I know, I don't see any issue here.

michaelDCurran commented:
@beweedon  I have managed to add functional tests for useCachedApi.
I have also added functional tests for ignoreDefault. I forgot that rather than raising an exception GetPropertyValue just returns a variant set to UI Automation's NotSupported value.
As a part of this work however, I changed UiaVariant's IsNotsupported and IsMixedAttribute methods to actually make use of the IUIAutomation instance passed in on initialize, and fetch ReservedNotSupportedValue and ReservedMixedAttributeValue from that, rather than taking those objects as arguments. Note that there was a comment above those methods stating it needed to be taken as an argument as this library had no access to the client API, which has always been untrue.
 This is an API breaking change though, in that the signatures for IsNotSupported and IsMixedAttribute have now changed, and it does rely on the fact that the caller initializes this library giving it a valid IUIAutomation instance, and not say just a nulptr.
I can write the ignoreDefault functional tests so that they manually get the IUIAutomation instance and pass it to IsNotSupported, and then revert the change, though for a long time now, I have wondered why IsNotSupported and IsMixedAttribute made that incorrect assumption... Is there something I'm missing, or is it perhaps just historical.

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+            // Tests that GetPropertyValue with ignoreDefault true raises an exception on an unimplemented property ID
+        // and that ignoreDefault false correctly returns a default value.
+        void GetPropertyValueIgnoreDefaultTest(const bool useRemoteOperations)
+                {
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+
+            UiaElement element = calc;
+
+            auto defaultVal = element.GetPropertyValue(UiaPropertyId(UIA_AriaPropertiesPropertyId), false);
```
Add comment to explain what `false` and `true` are for these calls. Example:

```cpp
auto defaultVal = element.GetPropertyValue(UiaPropertyId(UIA_AriaPropertiesPropertyId), false /*ignoreDefault*/);
```

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+        }
+
+        TEST_METHOD(GetPropertyValueIgnoreDefaultLocalTest)
+                {
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+        // Tests GetPropertyValue with useCachedAPI set to true and false
+        // Ensuring that an exception is raised if trying to fetch a cached value that is not cached. 
+        void GetPropertyValueUseCachedAPITest(const bool useRemoteOperations)
+                {
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+            cacheRequest.AddProperty(UIA_NamePropertyId);
+
+            auto elementWithCache = elementWithoutCache.GetUpdatedCacheElement(cacheRequest);
+scope.BindResult(elementWithCache);
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+            Assert::AreEqual(std::wstring(static_cast<wil::shared_bstr>(cachedName).get()), std::wstring(L"Display is 0"));
+
+            Assert::ExpectException<winrt::hresult_error>([&]()
+                {
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+        }
+
+        TEST_METHOD(GetPropertyValueUseCachedAPILocalTest)
+                {
```
**nit:** indentation

beweedon commented on src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp:
```diff
@@ -174,6 +174,121 @@ namespace UiaOperationAbstractionTests
+            });
+
+            UiaString uncachedName = elementWithoutCache.GetPropertyValue(UiaPropertyId(UIA_NamePropertyId), /*ignoreDefault=*/ false, /*useCachedApi=*/ false).AsString();
+            Assert::AreEqual(std::wstring(static_cast<wil::shared_bstr>(uncachedName).get()), std::wstring(L"Display is 0"));
```
For completeness, validate that we can use `useCachedApi = false` with `elementWithCache` as well.

beweedon commented on src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h:
```diff
@@ -1743,12 +1743,9 @@ namespace UiaOperationAbstraction
-        // It would be nice if this didn't have to be told what the unsupported attribute value is,
-        // but we don't have an IUIAutomation object handy in this library so it does.
-        UiaBool IsNotSupported(_In_ const IUnknown* unsupportedAttributeValue) const;
+        UiaBool IsNotSupported() const;
```
Hmm, I think you're right it doesn't make sense to need to take an `IUnknown` as a parameter when we do already have `g_automation`. Probably this is left over from a _very_ old version of the code. Let me think about it some more to make sure we're not missing anything.

beweedon replied:
OK, I've thought about this some. Yeah, this parameter here is absolutely something left over from a while ago. Good catch!

Personally, I think it's fine to in this change remove the parameter. This UiaOperationAbstraction is still super young and pre-versioned, so breaking API compatibility doesn't really bother me too much. Plus, this breaking compatibility would just result in a build break, so it shouldn't cause unexpected behavior if anyone were to update their UiaOperationAbstraction dependency. Does that seem reasonable from your perspective?

@pengyuanjin @zhayuli @mavangur any particular thoughts about this?

michaelDCurran replied:
I'm certainly good with keeping the change. It has confused me as to why it was there for a very long time :)

mlalic replied:
This was indeed a remnant from long ago - one of the first iterations of the library.

Originally, the idea was that this helper wouldn't maintain any state.

(In fact, at an even earlier point, the idea was to try to make this a header-only library; though that's orthogonal to whether or not you maintain static state, I believe there are still some remnants of that approach in other parts of the code that you might find, but I digress...)

Static libraries that do that are finnicky... What happens when you've got two clients in the same module (exe/dll) who might want to configure that static state differently? Maybe one uses Win7 IUIAutomation and the other the newer one... If you allow static state to be updated, then you've got to think about thread safety - which either leads to a "please don't do this concurrently" code comment which someone will not see and almost certainly eventually lead to a crash or a perf degradation due to too much locking. 

So, since we didn't want any static state, we wanted to avoid maintaining a `IUIAutomation` objects internally... As such, the only way to implement this would've been to pass the object at call time (either automation or the punk).

However, over time and once we started using this in real code, that proved overly limiting as more and more scenarios showed up where having that interface around would've been nifty and led to better code and interfaces. Meanwhile, there didn't seem to be a _real_ scenario where a single module (mostly DLLs) would want to initialize this differently. So we decided to adjust that original thinking and added the init function that accepts automation. The fact that it's "init" should hopefully also avert anyone from modifying it, mitigating the concurrency concerns. 

(I do think we left a warning comment somewhere about this potentially leading to surprising results if two different clients in the module run init?)

However, once we added that, we simply never went back to update these methods.

So +1 to this improvement. 

Hope the historic background helps.

michaelDCurran commented:
@beweedon Thanks for catching all the indenting issues there. It has motivated me to come up with a better way of reading them with my screen reader as a blind user :) Hopefully I have now addressed them all suitably.

beweedon commented:
> @beweedon Thanks for catching all the indenting issues there. It has motivated me to come up with a better way of reading them with my screen reader as a blind user :) Hopefully I have now addressed them all suitably.

No worries! Indentation is one of those things that only actually matters to a subset of programmers but it's good to do consistently anyway.

I think this looks good to me. I'll give a bit of time for any opinions on changing `IsNotSupported` etc. to filter in, then I'll approve.

beweedon commented:
@michaelDCurran permission to merge?

beweedon commented:
/azp run

michaelDCurran commented:
I give permission for this to be merged.
